---
title: "class09"
author: "Xinlong Wan"
format: pdf
---

# 1. Protein data bank

```{r}
pdb_df = read.csv("Data Export Summary.csv")
pdb_df
```

> Q1. 

```{r}
library(stringr)
pdb_df <- as.data.frame(lapply(pdb_df, str_remove, ","))
pdb_df[,-1] <- as.data.frame(lapply(pdb_df[,-1], strtoi))
```

```{r}
pdb_df
```


```{r}
print(paste(round(sum(pdb_df$X.ray)/sum(pdb_df$Total) * 100, 2), "% by X-Ray"))  

print(paste(round(sum(pdb_df$EM)/sum(pdb_df$Total) * 100, 2), "% by EM"))  
```

> Q2. 

```{r}
print(paste(round(sum(pdb_df[(pdb_df$Molecular.Type=="Protein (only)" | pdb_df$Molecular.Type=="Protein/Oligosaccharide"),"Total"])/sum(pdb_df$Total) *100,2), "%")) 
```

> Q3. 

Plain text search “HIV” doesn't give us a straight answer. We can maybe try restrict the search term to protease and hoping it catch all the answers.


# 2. Visualizing HIV-1 protease structure

> Q4.

Only the oxygen is visible, because the resolution is not enough for the hydrogen to be see. It's too small.

> Q5.

Choose "ligand", and we can find a water molecular in the center of the ligand. It's HOH 308, having 4 bonds that connect to both the ligand and the backbone chain.  

> Q6

!["HIV-Pr"](1HSG (3).png)

# 3. PDB data in R

> Q7. 198

> Q8. HOH

> Q9. 2

```{r}
library(bio3d)
pdb <- read.pdb("1hsg")
attributes(pdb)
```
```{r}
head(pdb$atom$resid[1])
```
Normal mode analysis (NMA) is a structural bioinformatics method to predict protein flexibility and potential functional motions (a.k.a. conformational changes).

```{r}
adk <- read.pdb("6s36")

# Perform flexiblity prediction
m <- nma(adk)
```

```{r}
plot(m)
```

```{r}
mktrj(m, file="adk_m7.pdb")
```

import this file into pdb and see the movement.

```{r}
# get the sequence by identifier
#get.seq()
#blast.pdb()
#get.pdb() 
#pdbaln() # align and superpose all structure

# pca()
# plot()
```


# Section 4/ Comparative structure analysis

> Q10. msa

> Q11. bio3d-view 

> 12. TRUE

```{r}
library(bio3d)
aa <- get.seq("1ake_A")
```

```{r}
aa
```
> Q13. 214

```{r}
# Blast or hmmer search 
#b <- blast.pdb(aa)
```

To render it without running blast each time but still have the "b" object, we can save the b object and load it next time. 

```{r}
# saveRDS(b, file = "blast_1ake_A.RDS")
b <- readRDS("blast_1ake_A.RDS")
```

```{r}
# Plot a summary of search results
hits <- plot(b)
```

```{r}
# List out some 'top hits'
head(hits$pdb.id)
```

```{r}
# if blast doesn't work, we can download the above ids directly from the database.

#hits <- NULL
#hits$pdb.id <- c('1AKE_A','6S36_A','6RZE_A','3HPR_A','1E4V_A','5EJE_A','1E4Y_A','3X2S_A','6HAP_A','6HAM_A','4K46_A','3GMT_A','4PZL_A')
```


```{r}
# Download releated PDB files
# the path argument make a new folder called "pdbs" and store the downloads in it.
# split by genes
# gzip make the file smaller
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

```{r}
# Align PDBs downloaded before
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")


```

```{r}
head(pdbs$id)
```


```{r}
# Vector containing PDB codes for figure axis, the basename function strip the id name out of the file path
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment

# plot(pdbs, labels=ids)
```

```{r}
# anotate collected PDB structures
# basically expand description of the ids
anno <- pdb.annotate(ids)
unique(anno$source)
```

```{r}
# pca

pc.xray <- pca(pdbs)
plot(pc.xray)
```

Each of the dots represent one PDB structure. 


PMSD is the pairwise RMSD values of the structural ensemble. It represent the "distance" of each pdb structure pair. It could be used as distance matrix for clustering. 

The clustering assigned a "label" to each class. It's sort of confirm the points that are clustered together in pca are indeed also close by rmsd distance. 

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)


# bg = plot background color
plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```

# Visualization

```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```


